File: server/Models/Ship.cs
using System.ComponentModel.DataAnnotations;

namespace server.Models;

public class Ship
{
    [Required]
    public long Id { get; set; }

    [Required]
    [StringLength(25, MinimumLength = 3, ErrorMessage = "The property Name should have a minimum length of 3 characters and not exceed 25 characters.")]
    public string Name { get; set; } = null!;

    [Required]
    [RegularExpression(@"^[A-Z]{4}-\d{4}-[A-Z]\d$",
         ErrorMessage = "The property Code should follow the required 'AAAA-1111-A1' pattern.")]
    public string Code { get; set; } = null!;

    [Required]
    public float Length { get; set; }

    [Required]
    public float Width { get; set; }
}

File: client/src/schemas/ship.ts
import { z } from "zod";

export const shipSchema = z.object({
  id: z.number().gt(0).optional(),
  name: z.string().min(3).max(25),
  length: z.number().gt(0),
  width: z.number().gt(0),
  code: z
    .string()
    .regex(
      new RegExp(/^[A-Z]{4}-\d{4}-[A-Z]\d$/),
      "Name should follow the pattern AAAA-1111-A1"
    ),
});
